pandas/core/dtypes/concat.py

```diff
@@ -559,3 +562,43 @@ def convert_sparse(x, axis):
         # coerce to object if needed
         result = result.astype('object')
     return result
+
+
+def _concat_indexes_same_dtype_rangeindex(indexes):
+    # Concatenates multiple RangeIndex instances. All members of "indexes" must
+    # be of type RangeIndex; result will be RangeIndex if possible, Int64Index
+    # otherwise. E.g.:
+    # indexes = [RangeIndex(3), RangeIndex(3, 6)] -> RangeIndex(6)
+    # indexes = [RangeIndex(3), RangeIndex(4, 6)] -> Int64Index([0,1,2,4,5])
+
+    start = step = next = None
+
+    for obj in indexes:
+        if not len(obj):
+            continue
+
+        if start is None:
+            # This is set by the first non-empty index
+            start = obj._start
+            if step is None and len(obj) > 1:
+                step = obj._step
+        elif step is None:
+            # First non-empty index had only one element
+            if obj._start == start:
+                return _concat_index_asobject(indexes)
+            step = obj._start - start
+
+        non_consecutive = ((step != obj._step and len(obj) > 1) or
+                           (next is not None and obj._start != next))
+        if non_consecutive:
+            # Not nice... but currently what happens in NumericIndex:
```

Comment:
I see, OK to keep it then, but maybe make it a bit more informative (or remove the 'not nice ', just say that it is what is used by Int64Index._append_same_dtype)

(I also don't think it would give that much of gain for making a special cased one of integers)